fix(metadata-store): #916 Ensure drop transaction gets committed

ncloudioj opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/activity-stream/917)
<!-- Reviewable:end -->

ncloudioj commented:
@oyiptong r?

@sarracini and @k88hudson have found the `asycnDrop` was a bit annoying if it couldn't guarantee the drop transaction gets committed. So I made a workaround to make it easier to use.

oyiptong commented on lib/MetadataStore.js:
```diff
@@ -280,6 +281,38 @@ MetadataStore.prototype = {
+   * the uncommitted state despite its promise is resolved in Sqlite.jsm.
+   * Hence, it has to poll for the table info periodically to comfirm that
+   */
+  _waitForDrop() {
```
please rename, it's not waiting for a drop, it's waiting for a reset

oyiptong commented:
from our conversation IRL: upon uninstall/disable, it is preferable to delete the database rather than reset the database, otherwise we'd leave an empty (but fully created) sqlite database  behind

ncloudioj commented:
Very good points. Will make those changes accordingly.

ncloudioj commented:
@sarracini FYI, I've refactored the MetadataStore a bit, specifically,
* rename `asyncDrop` to `asyncReset`, it resets all the tables (essentially, it does a combo - "drop" and then "create tables"
* add `asyncDelete` to nuke the SQLite file, I've also updated the code that handles the "disable/uninstall" events in `main.js`

ncloudioj commented:
@oyiptong One more round, please? :)

oyiptong commented on lib/MetadataStore.js:
```diff
@@ -280,21 +282,70 @@ MetadataStore.prototype = {
+   * the uncommitted state despite its promise is resolved in Sqlite.jsm.
+   * Hence, it has to poll for the table info periodically to comfirm that
+   */
+  _waitForAsyncReset() {
```
makes sense to move this back to the `test/lib/utils.js` now right?

ncloudioj replied:
Sure, will do.

oyiptong commented:
r+ with a nit: please move the `waitForReset` function back to `test/lib/utils.js`
